File: clients/clojure/examples/basic_usage.clj
(ns pg-mentat.examples.basic-usage
  "Basic usage examples for the pg_mentat Clojure client.

   This demonstrates the Datomic-compatible API: client creation,
   schema definition, transactions, queries, pull, and time travel.

   Prerequisites:
     - mentatd running at ws://localhost:8080/ws
     - A PostgreSQL database with pg_mentat extension installed

   Run:
     clj -M -m pg-mentat.examples.basic-usage"
  (:require [pg-mentat.client :as d]))

;; ---------------------------------------------------------------------------
;; 1. Connect to pg_mentat via mentatd
;; ---------------------------------------------------------------------------

(defn connect-example []
  (println "\n--- 1. Connecting ---")

  ;; Create a client (analogous to datomic.client.api/client)
  (let [client (d/client {:server-type :pg-mentat
                           :endpoint "ws://localhost:8080/ws"})]

    ;; List available databases
    (println "Available databases:" (d/list-databases client))

    ;; Connect to a specific database
    (let [conn (d/connect client {:db-name "example-db"})]
      (println "Connected to:" (:db-name conn))
      (println "Connection ID:" (:connection-id conn))
      conn)))

;; ---------------------------------------------------------------------------
;; 2. Define a schema
;; ---------------------------------------------------------------------------

(defn schema-example [conn]
  (println "\n--- 2. Schema Definition ---")

  ;; Schema attributes use the same EDN format as Datomic.
  ;; All standard value types are supported: string, long, double,
  ;; boolean, instant, keyword, ref, uuid, bytes.
  (d/transact conn
    {:tx-data
     [;; Person attributes
      {:db/ident       :person/name
       :db/valueType   :db.type/string
       :db/cardinality :db.cardinality/one
       :db/doc         "A person's full name"}

      {:db/ident       :person/email
       :db/valueType   :db.type/string
       :db/cardinality :db.cardinality/one
       :db/unique      :db.unique/identity
       :db/doc         "A person's email (unique identity)"}

      {:db/ident       :person/age
       :db/valueType   :db.type/long
       :db/cardinality :db.cardinality/one}

      {:db/ident       :person/roles
       :db/valueType   :db.type/keyword
       :db/cardinality :db.cardinality/many
       :db/doc         "Set of role keywords"}

      {:db/ident       :person/manager
       :db/valueType   :db.type/ref
       :db/cardinality :db.cardinality/one
       :db/doc         "Reference to this person's manager"}

      ;; Enum values (defined as named entities, just like Datomic)
      {:db/ident :role/engineer}
      {:db/ident :role/manager}
      {:db/ident :role/director}]})

  (println "Schema transacted successfully."))

;; ---------------------------------------------------------------------------
;; 3. Transact data
;; ---------------------------------------------------------------------------

(defn transact-example [conn]
  (println "\n--- 3. Transacting Data ---")

  ;; Map-form transactions (most common)
  (let [result (d/transact conn
                 {:tx-data
                  [{:db/id       "bob-id"      ;; string tempid
                    :person/name  "Bob Smith"
                    :person/email "bob@example.com"
                    :person/age   42
                    :person/roles [:role/manager]}

                   {:db/id        "alice-id"
                    :person/name  "Alice Johnson"
                    :person/email "alice@example.com"
                    :person/age   35
                    :person/roles [:role/engineer]
                    :person/manager "bob-id"}  ;; reference by tempid

                   {:person/name  "Carol Williams"
                    :person/email "carol@example.com"
                    :person/age   28
                    :person/roles [:role/engineer]
                    :person/manager "bob-id"}]})]

    (println "Transaction result keys:" (keys result))
    (println "Tempid mappings:" (:tempids result))
    result))

;; ---------------------------------------------------------------------------
;; 4. Query data
;; ---------------------------------------------------------------------------

(defn query-examples [conn]
  (println "\n--- 4. Queries ---")
  (let [db (d/db conn)]

    ;; Simple query: find all names
    (println "\nAll people:")
    (let [results (d/q '[:find ?name
                         :where [?e :person/name ?name]]
                       db)]
      (doseq [[name] results]
        (println "  -" name)))

    ;; Query with joins
    (println "\nPeople with emails:")
    (let [results (d/q '[:find ?name ?email
                         :where
                         [?e :person/name ?name]
                         [?e :person/email ?email]]
                       db)]
      (doseq [[name email] results]
        (println "  -" name "-" email)))

    ;; Query with input parameter
    (println "\nPeople older than 30:")
    (let [results (d/q '[:find ?name ?age
                         :in $ ?min-age
                         :where
                         [?e :person/name ?name]
                         [?e :person/age ?age]
                         [(>= ?age ?min-age)]]
                       db 30)]
      (doseq [[name age] results]
        (println "  -" name "(age" (str age ")"))))

    ;; Aggregate query
    (println "\nAge statistics:")
    (let [results (d/q '[:find (count ?e) (avg ?age) (min ?age) (max ?age)
                         :where
                         [?e :person/age ?age]]
                       db)]
      (let [[cnt avg-age min-age max-age] (first results)]
        (println "  Count:" cnt
                 "Average:" avg-age
                 "Min:" min-age
                 "Max:" max-age)))

    ;; Find spec: collection (returns flat list)
    (println "\nAll names (collection):")
    (let [results (d/q '[:find [?name ...]
                         :where [_ :person/name ?name]]
                       db)]
      (println "  " results))

    ;; Find spec: scalar (returns single value)
    (println "\nFirst name (scalar):")
    (let [result (d/q '[:find ?name .
                        :where [_ :person/name ?name]]
                      db)]
      (println "  " result))))

;; ---------------------------------------------------------------------------
;; 5. Pull API
;; ---------------------------------------------------------------------------

(defn pull-examples [conn]
  (println "\n--- 5. Pull API ---")
  (let [db (d/db conn)
        ;; First, find Alice's entity ID
        alice-id (d/q '[:find ?e .
                        :where [?e :person/email "alice@example.com"]]
                      db)]

    (when alice-id
      ;; Pull specific attributes
      (println "\nAlice (specific attrs):")
      (println "  " (d/pull db '[:person/name :person/email :person/age] alice-id))

      ;; Pull with wildcard
      (println "\nAlice (all attrs):")
      (println "  " (d/pull db '[*] alice-id))

      ;; Pull with nested reference (resolve manager)
      (println "\nAlice with manager details:")
      (println "  " (d/pull db '[:person/name
                                  {:person/manager [:person/name :person/email]}]
                            alice-id))

      ;; Pull with reverse lookup (who has Alice as manager?)
      (println "\nBob's direct reports:")
      (let [bob-id (d/q '[:find ?e .
                          :where [?e :person/email "bob@example.com"]]
                        db)]
        (when bob-id
          (println "  " (d/pull db '[:person/name
                                      {:person/_manager [:person/name]}]
                                bob-id)))))))

;; ---------------------------------------------------------------------------
;; 6. Time travel
;; ---------------------------------------------------------------------------

(defn time-travel-example [conn]
  (println "\n--- 6. Time Travel ---")
  (let [db (d/db conn)]

    ;; Record the current transaction ID
    (let [current-t (:t db)]
      (println "Current t:" current-t)

      ;; Make a change
      (d/transact conn {:tx-data [[:db/add
                                    (d/q '[:find ?e .
                                           :where [?e :person/email "alice@example.com"]]
                                         db)
                                    :person/age 36]]})

      ;; Query the current state
      (let [new-db (d/db conn)]
        (println "\nAlice's current age:"
                 (d/q '[:find ?age .
                        :where
                        [?e :person/email "alice@example.com"]
                        [?e :person/age ?age]]
                      new-db))

        ;; as-of: see the database before the change
        (let [old-db (d/as-of new-db current-t)]
          (println "Alice's age as-of t" (str current-t ":")
                   (d/q '[:find ?age .
                          :where
                          [?e :person/email "alice@example.com"]
                          [?e :person/age ?age]]
                        old-db)))

        ;; history: see all values over time
        (println "\nAlice's age history:")
        (let [hist-db (d/history new-db)
              results (d/q '[:find ?age ?tx ?added
                             :where
                             [?e :person/email "alice@example.com"]
                             [?e :person/age ?age ?tx ?added]]
                           hist-db)]
          (doseq [[age tx added] (sort-by second results)]
            (println "  t=" tx "age=" age (if added "asserted" "retracted"))))))))

;; ---------------------------------------------------------------------------
;; 7. Datoms and index access
;; ---------------------------------------------------------------------------

(defn datoms-example [conn]
  (println "\n--- 7. Datoms Index Access ---")
  (let [db (d/db conn)]

    ;; Access EAVT index
    (println "EAVT datoms (first 5):")
    (let [datoms (d/datoms db {:index :eavt})]
      (doseq [d (take 5 datoms)]
        (println "  " d)))

    ;; Access AEVT index (by attribute)
    (println "\nAEVT datoms for :person/name:")
    (let [datoms (d/datoms db {:index :aevt
                                :components [:person/name]})]
      (doseq [d datoms]
        (println "  " d)))))

;; ---------------------------------------------------------------------------
;; 8. Advanced query patterns
;; ---------------------------------------------------------------------------

(defn advanced-queries [conn]
  (println "\n--- 8. Advanced Query Patterns ---")
  (let [db (d/db conn)]

    ;; NOT clause: people without a manager
    (println "People without a manager:")
    (let [results (d/q '[:find ?name
                         :where
                         [?e :person/name ?name]
                         (not [?e :person/manager _])]
                       db)]
      (doseq [[name] results]
        (println "  -" name)))

    ;; OR clause: engineers or managers
    (println "\nEngineers or managers:")
    (let [results (d/q '[:find ?name ?role
                         :where
                         [?e :person/name ?name]
                         [?e :person/roles ?role]
                         (or [?e :person/roles :role/engineer]
                             [?e :person/roles :role/manager])]
                       db)]
      (doseq [[name role] results]
        (println "  -" name "(" role ")")))

    ;; Query with rules
    (println "\nUsing rules -- find all reports (direct + indirect):")
    (let [rules '[[(reports-to ?person ?manager)
                   [?person :person/manager ?manager]]
                  [(reports-to ?person ?manager)
                   [?person :person/manager ?middle]
                   (reports-to ?middle ?manager)]]
          results (d/q '[:find ?person-name ?manager-name
                         :in $ %
                         :where
                         (reports-to ?person ?manager)
                         [?person :person/name ?person-name]
                         [?manager :person/name ?manager-name]]
                       db rules)]
      (doseq [[person manager] results]
        (println "  " person "reports to" manager)))))

;; ---------------------------------------------------------------------------
;; 9. Connection cleanup
;; ---------------------------------------------------------------------------

(defn cleanup-example [conn]
  (println "\n--- 9. Cleanup ---")
  (d/release conn)
  (println "Connection released."))

;; ---------------------------------------------------------------------------
;; Main entry point
;; ---------------------------------------------------------------------------

(defn -main [& args]
  (println "=== pg_mentat Clojure Client -- Basic Usage ===")
  (println "Connecting to mentatd at ws://localhost:8080/ws")

  (try
    (let [conn (connect-example)]
      (schema-example conn)
      (transact-example conn)
      (query-examples conn)
      (pull-examples conn)
      (time-travel-example conn)
      (datoms-example conn)
      (advanced-queries conn)
      (cleanup-example conn))

    (catch Exception e
      (println "\nError:" (.getMessage e))
      (when-let [data (ex-data e)]
        (println "Details:" data))))

  (println "\n=== Done ===")
  (shutdown-agents))

File: clients/python/examples/basic_usage.py
#!/usr/bin/env python3
"""
Basic usage examples for the pg_mentat Python client.

This demonstrates both access paths:
  1. Direct PostgreSQL (recommended for new projects)
  2. Datomic-compatible API via mentatd WebSocket

Prerequisites:
  For direct PostgreSQL:
    pip install psycopg2-binary
    PostgreSQL running with pg_mentat extension

  For Datomic-compatible API:
    pip install websocket-client
    mentatd running at ws://localhost:8080/ws

Run:
    python basic_usage.py [--direct | --mentatd]
"""

import json
import sys

# ===========================================================================
# PATH 1: Direct PostgreSQL (recommended)
# ===========================================================================


def direct_postgresql_example():
    """
    Connect directly to PostgreSQL and call pg_mentat SQL functions.
    No mentatd daemon required. This is the recommended approach.
    """
    # Add parent directory to path for the direct client
    sys.path.insert(0, "..")
    from pg_mentat_client import MentatClient

    print("=== Direct PostgreSQL Access ===\n")

    # --- 1. Connect ---
    # Uses standard psycopg2 connection string.
    # Supports all psycopg2.connect() keyword arguments.
    with MentatClient("dbname=postgres host=localhost") as m:

        # --- 2. Define Schema ---
        print("--- Schema Definition ---")
        m.transact(
            """[
            {:db/ident :person/name
             :db/valueType :db.type/string
             :db/cardinality :db.cardinality/one
             :db/doc "A person's full name"}

            {:db/ident :person/email
             :db/valueType :db.type/string
             :db/cardinality :db.cardinality/one
             :db/unique :db.unique/identity}

            {:db/ident :person/age
             :db/valueType :db.type/long
             :db/cardinality :db.cardinality/one}

            {:db/ident :person/active
             :db/valueType :db.type/boolean
             :db/cardinality :db.cardinality/one}

            {:db/ident :person/friends
             :db/valueType :db.type/ref
             :db/cardinality :db.cardinality/many}
        ]"""
        )
        print("Schema created.\n")

        # --- 3. Transact Data ---
        print("--- Transacting Data ---")
        result = m.transact(
            """[
            {:db/id "alice"
             :person/name "Alice Johnson"
             :person/email "alice@example.com"
             :person/age 35
             :person/active true}

            {:db/id "bob"
             :person/name "Bob Smith"
             :person/email "bob@example.com"
             :person/age 42
             :person/active true
             :person/friends "alice"}

            {:person/name "Carol Williams"
             :person/email "carol@example.com"
             :person/age 28
             :person/active false}
        ]"""
        )
        print(f"Transaction result keys: {list(result.keys()) if isinstance(result, dict) else type(result)}\n")

        # --- 4. Query ---
        print("--- Queries ---")

        # Simple query
        print("\nAll people:")
        results = m.query("[:find ?name :where [?e :person/name ?name]]")
        for row in results:
            print(f"  - {row[0]}")

        # Query with joins
        print("\nPeople with emails:")
        results = m.query(
            """[:find ?name ?email
                :where
                [?e :person/name ?name]
                [?e :person/email ?email]]"""
        )
        for name, email in results:
            print(f"  - {name} <{email}>")

        # Query with input parameters (via JSON inputs)
        print("\nPeople older than 30:")
        results = m.query(
            """[:find ?name ?age
                :in $ ?min-age
                :where
                [?e :person/name ?name]
                [?e :person/age ?age]
                [(>= ?age ?min-age)]]""",
            inputs={"args": [30]},
        )
        for name, age in results:
            print(f"  - {name} (age {age})")

        # Aggregate query
        print("\nAge statistics:")
        results = m.query(
            """[:find (count ?e) (avg ?age) (min ?age) (max ?age)
                :where [?e :person/age ?age]]"""
        )
        if results:
            cnt, avg_age, min_age, max_age = results[0]
            print(f"  Count: {cnt}, Avg: {avg_age}, Min: {min_age}, Max: {max_age}")

        # --- 5. Pull ---
        print("\n--- Pull API ---")

        # Find Alice's entity ID first
        alice_results = m.query(
            '[:find ?e . :where [?e :person/email "alice@example.com"]]'
        )
        if alice_results:
            alice_id = alice_results
            print(f"\nAlice (entity {alice_id}):")
            entity = m.pull("[*]", alice_id)
            print(f"  {json.dumps(entity, indent=2, default=str)}")

        # --- 6. Entity ---
        print("\n--- Entity ---")
        if alice_results:
            entity = m.entity(alice_results)
            print(f"Alice entity: {json.dumps(entity, indent=2, default=str)}")

        # --- 7. Schema ---
        print("\n--- Schema ---")
        schema = m.schema()
        print(f"Schema type: {type(schema)}")
        if isinstance(schema, dict):
            print(f"Attributes: {list(schema.keys())[:5]}...")

        # --- 8. Explain ---
        print("\n--- Query Explain ---")
        plan = m.explain("[:find ?name :where [?e :person/name ?name]]")
        print(f"Execution plan: {json.dumps(plan, indent=2, default=str)[:200]}...")

        # --- 9. Time Travel (via inputs) ---
        print("\n--- Time Travel ---")

        # as-of query
        results = m.query(
            "[:find ?name :where [?e :person/name ?name]]",
            inputs={"asOf": 268435460},
        )
        print(f"Names as-of tx 268435460: {results}")

        # History query
        results = m.query(
            """[:find ?e ?name ?tx ?added
                :where [?e :person/name ?name ?tx ?added]]""",
            inputs={"history": True},
        )
        print(f"History rows: {len(results) if results else 0}")

        # --- 10. Native SQL View Access ---
        print("\n--- Native SQL Views ---")

        # Look up entities by attribute value
        alices = m.lookup_entity(":person/name", "Alice Johnson")
        print(f"Lookup 'Alice Johnson': {alices}")

        # Get a single value
        if alices:
            val = m.entity_value(alices[0]["entity_id"], ":person/email")
            print(f"Alice's email: {val}")

        # Full-text search
        print("\nFull-text search for 'alice':")
        text_results = m.find_text("alice")
        for r in text_results:
            print(f"  entity={r['entity_id']} attr={r['attribute']} rank={r['rank']}")

        # Browse facts
        if alices:
            print(f"\nFacts for entity {alices[0]['entity_id']}:")
            facts = m.facts(entity_id=alices[0]["entity_id"])
            for f in facts:
                print(f"  {f['attribute']} = {f['value']} ({f['value_type']})")

        # Transaction log
        print("\nRecent transactions:")
        txs = m.tx_log(limit=3)
        for tx in txs:
            print(f"  tx={tx['tx']} datoms={tx['datom_count']}")

    print("\n--- Connection auto-closed by context manager ---")


# ===========================================================================
# PATH 2: Datomic-Compatible API (via mentatd)
# ===========================================================================


def datomic_compatible_example():
    """
    Connect to mentatd via WebSocket using the Datomic-compatible API.
    Use this path when migrating from Datomic.
    """
    sys.path.insert(0, "..")
    import pg_mentat

    print("=== Datomic-Compatible API (via mentatd) ===\n")

    # --- 1. Create Client and Connect ---
    print("--- Connecting ---")
    c = pg_mentat.client(endpoint="ws://localhost:8080/ws")
    conn = pg_mentat.connect(c, db_name="example-db")
    print(f"Connected to: {conn.db_name}")

    try:
        # --- 2. Get Database Value ---
        database = pg_mentat.db(conn)
        print(f"Database t={database.t}")

        # --- 3. Schema ---
        print("\n--- Schema ---")
        pg_mentat.transact(
            conn,
            tx_data="""[
            {:db/ident :person/name
             :db/valueType :db.type/string
             :db/cardinality :db.cardinality/one}
            {:db/ident :person/email
             :db/valueType :db.type/string
             :db/cardinality :db.cardinality/one
             :db/unique :db.unique/identity}
            {:db/ident :person/age
             :db/valueType :db.type/long
             :db/cardinality :db.cardinality/one}
        ]""",
        )
        print("Schema transacted.")

        # --- 4. Transact Data ---
        print("\n--- Transacting ---")
        pg_mentat.transact(
            conn,
            tx_data="""[
            {:person/name "Alice" :person/email "alice@example.com" :person/age 35}
            {:person/name "Bob" :person/email "bob@example.com" :person/age 42}
        ]""",
        )
        print("Data transacted.")

        # --- 5. Query ---
        print("\n--- Query ---")
        database = pg_mentat.db(conn)  # refresh to see new data

        results = pg_mentat.q(
            '[:find ?name ?email :where [?e :person/name ?name] [?e :person/email ?email]]',
            database,
        )
        print(f"Results: {results}")

        # --- 6. Pull ---
        print("\n--- Pull ---")
        # Find entity ID first
        eid_result = pg_mentat.q(
            '[:find ?e . :where [?e :person/email "alice@example.com"]]',
            database,
        )
        if eid_result:
            entity = pg_mentat.pull(database, "[*]", eid_result)
            print(f"Alice: {entity}")

        # --- 7. Time Travel ---
        print("\n--- Time Travel ---")
        old_db = pg_mentat.as_of(database, database.t - 1)
        old_results = pg_mentat.q(
            "[:find ?name :where [?e :person/name ?name]]",
            old_db,
        )
        print(f"Names as-of t-1: {old_results}")

        # History
        hist_db = pg_mentat.history(database)
        hist_results = pg_mentat.q(
            "[:find ?e ?name ?tx ?added :where [?e :person/name ?name ?tx ?added]]",
            hist_db,
        )
        print(f"History entries: {len(hist_results) if hist_results else 0}")

        # --- 8. List Databases ---
        print("\n--- Catalog ---")
        dbs = pg_mentat.list_databases(c)
        print(f"Available databases: {dbs}")

    finally:
        # --- 9. Cleanup ---
        conn.close()
        print("\nConnection closed.")


# ===========================================================================
# Main
# ===========================================================================

if __name__ == "__main__":
    mode = sys.argv[1] if len(sys.argv) > 1 else "--direct"

    if mode == "--mentatd":
        datomic_compatible_example()
    elif mode == "--direct":
        direct_postgresql_example()
    else:
        print("Usage: python basic_usage.py [--direct | --mentatd]")
        print()
        print("  --direct   Use direct PostgreSQL access (default, recommended)")
        print("  --mentatd  Use Datomic-compatible API via mentatd WebSocket")
        sys.exit(1)
